Print current page option added to use current page option in Windows print dialog

sz5000 opened the pull request:
The pull request makes it possible to use the print current page option of the Windows print dialog.

![image](https://user-images.githubusercontent.com/33447587/224433417-04db63d2-be73-4ff4-a3d1-bd38b91b6318.png)

I have added three new members to wxPrintDialogData which can be used to control the new feature:
`int m_printCurrentPage` - the page number of the current page
`bool m_printEnableCurrentPage` - used to enable/disable the current page field in print dialog
`bool m_printSelectCurrentPage` - if set to true, the current page field is selected when the dialog is shown

By default the print current page option is disabled. Therefore the extension is 100% compatible with the old code.

This extension uses the new properties only for the Windows port. I don't know whether an option for a current page is also available in other dialogs so that these ones could be extended as well.

sz5000 committed:
Print current page option added to use current page option in Windows print dialog.

sz5000 committed:
Print current page option added to use current page option in Windows print dialog.

sz5000 committed:
whitespace fix

sz5000 committed:
Merge branch 'print_current_page' of https://github.com/sz5000/wxWidgets into print_current_page

vadz commented on src/msw/printdlg.cpp:
```diff
@@ -911,12 +927,19 @@ bool wxWindowsPrintDialog::ConvertToNative( wxPrintDialogData &data )
     if ( data.GetEnableHelp() )
         pd->Flags |= PD_SHOWHELP;
 
+#if(WINVER >= 0x0500)
```
We never use `WINVER` checks in wx code. This particular one is obsolete and can be simply removed anyhow because we only support Windows 7 or later (and probably won't support even Windows 7 for much longer) in master, but if you ever need to perform any checks, they need to be done during run-time and not compile-time because we want the same wx DLL to work on all Windows versions.

sz5000 replied:
Ok, changed.

vadz commented on src/msw/printdlg.cpp:
```diff
@@ -964,6 +991,11 @@ bool wxWindowsPrintDialog::ConvertFromNative( wxPrintDialogData &data )
     data.EnablePageNumbers( ((pd->Flags & PD_NOPAGENUMS) != PD_NOPAGENUMS) );
     data.EnableHelp( ((pd->Flags & PD_SHOWHELP) == PD_SHOWHELP) );
 
+#if(WINVER >= 0x0500)
```
Another check which should be removed.

sz5000 replied:
Changed.

vadz commented on src/msw/printdlg.cpp:
```diff
@@ -852,14 +857,16 @@ bool wxWindowsPrintDialog::ConvertToNative( wxPrintDialogData &data )
 
-    pd = new PRINTDLG;
-    memset( pd, 0, sizeof(PRINTDLG) );
+    pd = (LPPRINTDLGEX) GlobalAlloc(GPTR, sizeof(PRINTDLGEX));
```
I don't really mind doing this, but I just wonder why did you decide to change this, i.e. switch from `new` to `GlobalAlloc` here? There doesn't seem to be any reason to use the Windows function as we only free this one ourselves, so why?

sz5000 replied:
Ok, changed.

vadz commented on src/msw/printdlg.cpp:
```diff
@@ -964,6 +991,11 @@ bool wxWindowsPrintDialog::ConvertFromNative( wxPrintDialogData &data )
 
+#if(WINVER >= 0x0500)
+    data.SetSelectCurrentPage(((pd->Flags & PD_CURRENTPAGE) == PD_CURRENTPAGE));
+    data.EnableCurrentPage(((pd->Flags & PD_NOCURRENTPAGE) != PD_NOCURRENTPAGE));
```
I don't think this can be changed by user, can it? I.e. I think this line is just unnecessary.

sz5000 replied:
My last commit has changed this.

vadz commented on include/wx/cmndata.h:
```diff
@@ -170,20 +171,25 @@ class WXDLLIMPEXP_CORE wxPrintDialogData: public wxObject
     void EnableSelection(bool flag) { m_printEnableSelection = flag; }
     void EnablePageNumbers(bool flag) { m_printEnablePageNumbers = flag; }
     void EnableHelp(bool flag) { m_printEnableHelp = flag; }
+    void EnableCurrentPage(bool flag) { m_printEnableCurrentPage = flag; }
```
This is very minor, but I'd put this just below `EnableSelection()` because these functions are related (while it's completely unrelated to `EnableHelp()`).

I'd also consider renaming it to `EnableSelectCurrentPage()` to be consistent with `SetSelectCurrentPage()`. Alternatively, the other function could be renamed to just `SetCurrentPage()` but this would be less clear.

sz5000 replied:
Ok, changed.

vadz commented on include/wx/cmndata.h:
```diff
@@ -160,6 +160,7 @@ class WXDLLIMPEXP_CORE wxPrintDialogData: public wxObject
     int GetMinPage() const { return m_printMinPage; }
     int GetMaxPage() const { return m_printMaxPage; }
     int GetNoCopies() const { return m_printNoCopies; }
+    int GetCurrentPage() const { return m_printCurrentPage; }
```
This seems completely unused to me, as `m_printCurrentPage` is never set (nor used) anywhere. What is the point of having this function? IMO it should be just removed.

sz5000 replied:
Was not unsused, but my last commit has changed this too.

vadz commented on include/wx/cmndata.h:
```diff
@@ -170,20 +171,25 @@ class WXDLLIMPEXP_CORE wxPrintDialogData: public wxObject
     void SetMinPage(int v) { m_printMinPage = v; }
     void SetMaxPage(int v) { m_printMaxPage = v; }
     void SetNoCopies(int v) { m_printNoCopies = v; }
+    void SetCurrentPage(int v) { m_printCurrentPage = v; }
```
This should be removed too AFAICS.

sz5000 replied:
Changed with my last commit.

vadz commented on include/wx/cmndata.h:
```diff
@@ -170,20 +171,25 @@ class WXDLLIMPEXP_CORE wxPrintDialogData: public wxObject
     bool GetEnableSelection() const { return m_printEnableSelection; }
     bool GetEnablePageNumbers() const { return m_printEnablePageNumbers; }
     bool GetEnableHelp() const { return m_printEnableHelp; }
+    bool GetEnableCurrentPage() const { return m_printEnableCurrentPage; }
```
Similarly, I'd put this one right below `GetEnableSelection()`.

sz5000 replied:
Ok, changed.

vadz commented on include/wx/cmndata.h:
```diff
@@ -170,20 +171,25 @@ class WXDLLIMPEXP_CORE wxPrintDialogData: public wxObject
     bool GetEnablePageNumbers() const { return m_printEnablePageNumbers; }
     bool GetEnableHelp() const { return m_printEnableHelp; }
+    bool GetEnableCurrentPage() const { return m_printEnableCurrentPage; }
+    bool GetSelectCurrentPage() const { return m_printSelectCurrentPage; }
```
More importantly, I'd move this one to a different function group, together with `GetSelection()` -- as it's a completely kind of flag (this one is user input and not an option set by the program).

sz5000 replied:
Ok, changed

vadz commented on include/wx/cmndata.h:
```diff
@@ -209,6 +216,8 @@ class WXDLLIMPEXP_CORE wxPrintDialogData: public wxObject
     bool            m_printEnablePageNumbers;
     bool            m_printEnableHelp;
     bool            m_printEnablePrintToFile;
+    bool            m_printEnableCurrentPage;
```
Again, for consistency this one should be
```suggestion
    bool            m_printEnableSelectCurrentPage;
```

vadz reviewed (changes requested):
Thanks for your work on this, I agree that it would be definitely useful to have support for this, but I think the proposed API is a bit confusing and should be simplified/streamlined a bit.

It would also be very nice if you could please modify the printing sample to show the new API in action -- it's always the best way to see if it works well.

And, finally, once the API is stabilized, with all the functions names settled, we absolutely need documentation for it, so please update `interface/wx/cmndata.h` with the new functions descriptions. Please use `@since 3.3.0` in them.

TIA!

sz5000 committed:
Added function documentation for EnableCurrentPage, GetCurrentPage, SetCurrentPage

sz5000 committed:
Added function documentation for IsPageSelected and extended documentation for GetPageInfo to describe how to use the function arguments for current page and selection mode.

sz5000 committed:
Functions and names have been adapted to the previous names and the saving of the current page has been separated from wxPrintDialogData, so that the current page mode is structured like the selection mode.

sz5000 committed:
Added new function IsPageSelected(int page) needed to query selected pages. There used to be no way to do this, making selection mode pretty much useless.

sz5000 committed:
Removed GlobalAlloc and replaced it with new.

Minor changes to the structure to better identify related blocks.

Fixed problems because some combinations of PD_X flags don't go together, e.g. PD_SELECTION and PD_NOSELECTION.

sz5000 committed:
Because the selection mode wasn't really usable, it had to be improved here.

If GetAllPages():
  minPageNum = minPage and maxPageNum = maxPage
If GetSelection():
  minPageNum = fromPage and maxPageNum = toPage queried from printOut
If GetCurrentPage():
  minPageNum = fromPage and maxPageNum = toPage queried from printOut and fromPage and toPage should be equal

Furthermore, in order to only print selected pages in selection mode and to support non-consecutive pages, there is a new check inside the print loop which calls IsPageSelected from printOut to check if a page is selected. If not, this page is skipped.

vadz reviewed (commented):
Thanks a lot for the changes, I think the "current page" part is perfectly fine now, but it looks like there is also a new change, related to `IsPageSelected()` addition and now I don't understand how does this work... Who is supposed to override it? Shouldn't this be done by MSW printout using the information about multiple ranges from the dialog?

Again, showing the new functionality in the sample would probably be very useful to demonstrate how it works.

Also, could we perhaps leave this for another PR and merge just the current-page-related parts of this one?

TIA!

sz5000 commented:
"Also, could we perhaps leave this for another PR and merge just the current-page-related parts of this one?"
Yes

IsPageSelected() is required for the selection mode, otherwise the print function does not know which page is selected because currently there is function to get this information. I changed the msw/printWin.cpp function and there IsPageSelected() is used. I would definitively recommend to improve this because currently the print selection mode is pretty useless.

```
if ( !(m_printDialogData.GetAllPages() || m_printDialogData.GetSelection()) )
{
	minPageNum = m_printDialogData.GetFromPage();
	maxPageNum = m_printDialogData.GetToPage();
}
```

minPageNum and maxPageNum is only updated if the print range mode is used. So in selection mode, every page is printed because there is no code that determines selected pages. Therefore the new IsPageSelected() and this new method fixes this. It should be overridden by a PrintOut if selection mode is enabled.

vadz commented:
I don't understand how is this fix sufficient. I.e. who is supposed to override `IsPageSelected()` and how? It seems to me that this part is still missing, but perhaps it's me who's missing something, could you please explain me how is it supposed to work? Again, an example of using this in the sample would probably do wonders.

But if it's really not fully implemented yet, I'd rather avoid committing incomplete fixes. I can exclude them manually when merging if you're ok with this.

sz5000 commented:
Actually it's really simple. User clicks the pint function and selects the Selection radio button. Then the printing code needs some information from the app on which pages are selected so that only these pages are printed. This part is missing currently. In order so solve this, IsPageSelected() comes into action. If print selection is used, the printing code calls IsPageSelected with the page number and the printout has to respond either with true or false. If the return value it true, then the page is printed. If false is returned, then the page is skipped.

```
class MyPrintOut: public wxPrintout {
...
public:
	bool IsPageSelected(int page) {
		// return true if the page is selected in the APP/GUI, else flase
	}
...
};
```

This new method is used in printwin.cpp, line 207:

```
if (m_printDialogData.GetSelection() && !printout->IsPageSelected(pn))
                continue;
```

In wxWindowsPrinter class, if a document is printed, the implementation loops over the pages (minPageNum, maxPageNum) and prints the pages in that range but it does not check if a page in that range is selected or not. The two code lines above check that and skips pages that are not selected. How pages are selected is completely app specific.

sz5000 committed:
PrintDlgEx returns E_INVALIDARG if nFromPage is greater than nToPage

sz5000 committed:
printing sample updated to see the new current page feature and IsPageSelected() function in action

sz5000 commented:
I have also updated the printing sample which shows how to use the new current page feature and how to use the new IsPageSelected() function.

sz5000 committed:
psprint fix

sz5000 committed:
this hopefully fixes mac and linux build errors

vadz commented on interface/wx/print.h:
```diff
@@ -705,6 +705,13 @@ class wxPrintout : public wxObject
         and maximum page values that the user can select, and the required page range to
         be printed.
 
+        If the user has decided to print only selected pages, then pageFrom and pageTo are
```
```suggestion
        If the user has decided to print only selected pages, then @a pageFrom and @a pageTo are
```

vadz commented on interface/wx/print.h:
```diff
@@ -705,6 +705,13 @@ class wxPrintout : public wxObject
+        used to limit the page range and IsPageSelected() is used to query whether the page
+        is selected.
+
+        If the user only wants to print the current page, then pageFrom and pageTo are used
+        to set the current page by setting pageFrom = pageTo.
```
```suggestion
        If the user only wants to print the current page, then @a pageFrom and @a pageTo are both set to the current page number.
```

vadz reviewed (commented):
Thanks, this is much more clear now and, except for some minor docs improvements, I think is ready to be merged.

I'm going to separate it in two commits, one adding current page selection and the other one adding `IsPageSelected()`, just to keep things clean, when merging.

Please let me know if you plan to make any more changes here, otherwise I'll try to do it (probably towards the end of the week).

Thanks again!

sz5000 committed:
Update interface/wx/print.h

Co-authored-by: VZ <vz-github@zeitlins.org>

sz5000 committed:
Update interface/wx/print.h

Co-authored-by: VZ <vz-github@zeitlins.org>

sz5000 committed:
Update interface/wx/cmndata.h

Co-authored-by: VZ <vz-github@zeitlins.org>

sz5000 committed:
Update interface/wx/cmndata.h

Co-authored-by: VZ <vz-github@zeitlins.org>

sz5000 commented on interface/wx/cmndata.h:
```diff
@@ -489,6 +489,13 @@ class wxPrintDialogData : public wxObject
 
+    /**
+        Enables or disables the "Current Page" radio button.
+
```
Sounds reasonable if other print dialogs don't have a Current Page option.

sz5000 commented on include/wx/prntbase.h:
```diff
@@ -278,6 +278,7 @@ class WXDLLIMPEXP_CORE wxPrintout: public wxObject
     virtual bool HasPage(int page);
     virtual bool OnPrintPage(int page) = 0;
     virtual void GetPageInfo(int *minPage, int *maxPage, int *pageFrom, int *pageTo);
+    virtual bool IsPageSelected(int page);
```
I also don't think IsPageSelected changes the Object. So const probably wouldn't be a problem. But in the end it depends more on how the app implements this function and I'm just wondering if we're taking too much liberty with const for a programmer.

sz5000 committed:
trailing whitespace removed

sz5000 commented:
I am happy with the current state and I don't see any more problems. So, please merge.

vadz commented:
Sorry for the delay, I've finally repackaged your changes in 2 separate commits in the linked PR (e93570f681 and 498ed666da). The PR also contains some cleanup which I did after looking at this code, but the commits above contain just your changes with some minor tweaks, mostly to avoid gratuitous differences with the existing code.

Please let me know if you see any problems with the other PR, TIA!

P.S. I'm closing this one as it's a subset of the other one.
